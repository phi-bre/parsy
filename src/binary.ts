import {ParsyContext, ParsyParser} from '.';

export abstract class BinaryParser extends ParsyParser {
    public left!: ParsyParser;
    public right!: ParsyParser;

    public set(left: ParsyParser, right: ParsyParser): this {
        this.left = left;
        this.right = right;
        return this;
    }
}

export class AndParser extends BinaryParser {
    public parse(context: ParsyContext): ParsyContext | undefined {
        return this.left.parse(context) && this.right.parse(context);
    }
}

export class OrParser extends BinaryParser {
    public parse(context: ParsyContext): ParsyContext | undefined {
        return this.left.parse(context) || this.right.parse(context);
    }
}
