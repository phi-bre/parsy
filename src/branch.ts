import {Token} from './token';
import {indent} from './utils';

export class Branch extends Array<Branch | Token> {

    constructor(public type: string | number | symbol, public parent?: Branch) {
        super();
    }

    toString(level: number = 0): string {
        const space = indent(level);

        let children = '';
        for (const child of this) {
            children += child.toString(level + 1);
        }

        return space
            + this.type.toString()
            + ' [\n'
            + children
            + space
            + ']\n';
    }
}
