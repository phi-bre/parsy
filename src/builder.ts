import {BuilderOptions, Matcher, Node, State, Tokens} from './index';

const matchers = {};

export function builder(options: BuilderOptions) {
    const state = {matchers} as State;

    if (!options.scopes)
        throw 'No scopes specified.';
    if (!options.rules)
        throw 'No rules specified.';
    if (!options.start)
        throw 'No start specified.';

    convert(state, options.terminals, type => state => {
        const token = state.tokens.next;
        if (token.type === type) {
            return state.node.push(token);
        }
        throw 'Unexpected token: ' + token.type.toString() + '. \nExpected: ' + type;
    });

    convert(state, options.rules, type => state => {
        resolve(options.rules[type], state);
    });

    convert(state, options.scopes, type => state => {
        try {
            state.node = new Node(type, state.node);
            resolve(options.scopes[type], state);
            state.node.parent!.push(state.node);
        } finally {
            state.node = state.node.parent!;
        }
    });

    return function (tokens: Tokens) {
        const root = new Node('root');
        state.node = root;
        state.tokens = tokens;
        resolve(options.start, state);

        if (state.tokens.peek)
            throw 'Unexpected token: ' + state.tokens.next.type.toString() + '\nExpected: EOF';

        return root;
    }
}

/**
 * Used internally to convert config types to matchers.
 * @param state
 * @param types
 * @param reducer
 */
function convert(state: State, types, reducer: (type: string) => Matcher) {
    for (const type in types) {
        if (types.hasOwnProperty(type)) {
            if (type in state.matchers)
                throw 'Token type already defined: ' + type;
            state.matchers[type] = reducer(type);
        }
    }
}

/**
 * Used to register new matchers factories.
 * @param types
 * @param label
 * @param matcher
 */
export function register(types: (string | symbol)[], label: string, matcher: Matcher): symbol {
    if (!types?.length)
        throw 'A matcher must have at least one rule.';

    const symbol = Symbol(label);
    matchers[symbol] = matcher;
    return symbol;
}

/**
 * Used to resolve match type references to actual matchers.
 * @param type
 * @param state
 */
export function resolve(type: string | symbol, state: State) {
    if (!state.matchers[type as string])
        throw 'Cannot find reference: ' + type.toString();
    return state.matchers[type as string](state);
}

/**
 * Matches a sequence of types.
 * All the rules provided need to match.
 */
export function sequence(...types: (string | symbol)[]) {
    return register(types, 'sequence', state => {
        for (const type of types) {
            resolve(type, state);
        }
    });
}

/**
 * Searches for the first type that matches in sequential order.
 * At least one of the types has to match.
 */
export function alternation(...types: (string | symbol)[]) {
    return register(types, 'alternation', state => {
        const position = state.tokens.position;
        for (const type of types) try {
            return resolve(type, state);
        } catch {
            state.tokens.reset(position);
        }
        throw 'Unexpected token: ' + state.tokens.peek!.type.toString() + '\nExpected to find: ' + types.join(' | ');
    });
}

/**
 * Matches the supplied sequence but recovers if not all of them match.
 * Makes the sequence optional.
 */
export function optional(...types: (string | symbol)[]) {
    return register(types, 'optional', state => {
        const position = state.tokens.position;
        try {
            for (const type of types) {
                resolve(type, state);
            }
        } catch {
            state.tokens.reset(position);
        }
    });
}

/**
 * Matches and repeats matching the supplied sequence.
 * Recovers if the sequence never matches.
 */
export function star(...rules: (string | symbol)[]) {
    const reference = sequence(...rules);
    return register(rules, 'star', state => {
        while (true) {
            const position = state.tokens.position;
            try {
                resolve(reference, state);
            } catch {
                state.tokens.reset(position);
                break;
            }
        }
    });
}

/**
 * Matches and repeats matching the supplied sequence.
 * Requires the sequence to match at least once.
 */
export function plus(...types: (string | symbol)[]) {
    const reference = star(...types);
    return register(types, 'plus', state => {
        resolve(reference, state);
        // TODO
    });
}
