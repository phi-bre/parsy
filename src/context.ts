import {ParsyToken} from '.';

export class ParsyContext {
    public input: string;
    public index: number;
    public token!: ParsyToken;

    constructor(input: string, index: number = 0) {
        this.input = input;
        this.index = index;
        this.token = new ParsyToken(index);
        this.token.to = index;
    }

    public advance(length: number): this {
        this.index += length;
        return this;
    }

    public open(type?: string): this {
        this.token = new ParsyToken(this.index, this.token, type);
        return this;
    }

    public close(type?: string): this {
        if (!this.token.parent) throw 'Scope mismatch';
        this.token.value = this.input.substring(this.token.from, this.index);
        this.token.to = this.index;
        this.token.parent.children.push(this.token);
        this.token = this.token.parent;
        return this;
    }

    public discard(): this {
        if (!this.token.parent) throw 'Scope mismatch';
        this.token = this.token.parent;
        return this;
    }
}
