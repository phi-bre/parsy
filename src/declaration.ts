import {Node} from './node';
import {Parser} from './parser';

export abstract class Declaration {
    constructor(public type: string | symbol) {
    }

    public abstract parse(parser: Parser): Node | undefined | void;
}
