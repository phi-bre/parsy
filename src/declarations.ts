import {Rule} from './parser';

/**
 * Matches a sequence of rules.
 * All the rules provided need to match.
 */
export function sequence(...rules: (string | Rule)[]): Rule {
    return function (parser) {
        for (const rule of rules) {
            parser.resolve(rule);
        }
    }
}

/**
 * Searches for the first type that matches in sequential order.
 * At least one of the rules has to match.
 */
export function alternation(...rules: (string | Rule)[]): Rule {
    return function (parser) {
        const position = parser.lexer.position;
        for (const rule of rules) {
            try {
                parser.resolve(rule);
                return;
            } catch {
                parser.lexer.reset(position);
            }
        }
        throw `Unexpected token: ${parser.lexer.peek!.type.toString()}. Expected to find: ${rules.join(', ')}`;
    }
}

/**
 * Matches the supplied sequence but recovers if not all of the rules match.
 * Makes the sequence optional.
 */
export function optional(...rules: (string | Rule)[]): Rule {
    const seq = sequence(...rules);
    return function (parser) {
        const position = parser.lexer.position;
        try {
            seq(parser);
        } catch {
            parser.lexer.reset(position);
            return false;
        }
    }
}

/**
 * Matches and repeats matching the supplied sequence.
 * Recovers if the sequence never matches.
 */
export function star(...rules: (string | Rule)[]): Rule {
    const seq = optional(sequence(...rules));
    return function (parser) {
        while (seq(parser) !== false) ;
    }
}

/**
 * Matches and repeats matching the supplied sequence.
 * Requires the sequence to match at least once.
 */
export function plus(...rules: (string | Rule)[]): Rule {
    const seq = sequence(...rules);
    const opt = optional(seq);
    return function (parser) {
        seq(parser);
        while (opt(parser) !== false) ;
    }
}
