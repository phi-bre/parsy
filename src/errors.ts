import {Position, Token} from './';
import {indent} from './utils';

function report(input: string, {index, line, column}: Position) {
    const info = '\nindex: ' + index + ' line: ' + line + ' column: ' + column;
    let from = input.substring(0, index).lastIndexOf('\n');
    let to = input.indexOf('\n', index);
    if (from === -1) from = 0;
    if (to === -1) to = input.length;
    const subject = '\n' + input.substring(from, to);
    const underline = '\n' + indent(index - 1, ' ') + '~~~'; // TODO
    return info + subject + underline;
}

export class TokenError extends Error {
    public severity = 1;

    constructor(input: string, position: Position) {
        super(
            'Unexpected token: "'
            + input.substr(position.index, 1)
            + '"'
            + report(input, position)
        );
    }
}

export class SyntaxError extends Error {
    public severity = 1;

    constructor(input: string, token: Token, expected: string) {
        super(
            'Unexpected token: "'
            + token.value
            + '". expected: '
            + expected
            + report(input, token.position)
        );
    }
}
