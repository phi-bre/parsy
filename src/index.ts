import {Branch} from './branch';
import {Lexer, Terminal} from './lexer';
import {Parser, Rule, RuleStore} from './parser';
import {Token} from './token';

export * from './declarations';
export * from './parser';
export * from './lexer';
export * from './branch';
export * from './token';

export interface Position {
    index: number;
    line: number;
    column: number;
}

export function parsy(config) {
    const rules: RuleStore = {};
    const terminals: Terminal[] = [];

    if (config.ignore) {
        const symbol = Symbol('ignore');
        rules[symbol as unknown as string] = advance(symbol as unknown as string);
        const terminal = typeof config.terminals[symbol] === 'string' ? string : regex;
        terminals.push(terminal(symbol, config.ignore, true));
    }

    for (const identifier in config.terminals) if (config.terminals.hasOwnProperty(identifier)) {
        rules[identifier] = advance(identifier);
        const terminal = typeof config.terminals[identifier] === 'string' ? string : regex;
        terminals.push(terminal(identifier, config.terminals[identifier]));
    }

    for (const identifier in config.rules) if (config.rules.hasOwnProperty(identifier)) {
        if (identifier in rules)
            throw 'Terminals and rules cannot have the same identifier: ' + identifier;
        rules[identifier] = rule(identifier, config.rules[identifier]);
    }

    return function (input: string) {
        const lexer = new Lexer(terminals, input);
        const parser = new Parser(config.start, rules, lexer);
        return parser.tree;
    }
}

function advance(identifier: string): Rule {
    return function ({lexer, node}) {
        const token = lexer.next;
        if (token.type !== identifier)
            throw 'Unexpected token: ' + token.type.toString() + '. \nExpected: ' + identifier;
        node!.push(token);
    }
}

function rule(identifier: string, rule: Rule): Rule {
    return function (parser) {
        try {
            parser.node = new Branch(identifier, parser.node);
            rule(parser);
            parser.node!.parent!.push(parser.node);
        } finally {
            parser.node = parser.node!.parent;
        }
    }
}

function string(type: string | symbol, value: string, ignore?: boolean): Terminal {
    return function ({input, index, position}: Lexer) {
        if (input.startsWith(value, index)) {
            return new Token(type, value, position, ignore);
        }
    }
}

function regex(type: string | symbol, value: RegExp, ignore?: boolean): Terminal {
    const pattern = new RegExp(value.source, 'y');
    return function ({input, index, position}: Lexer) {
        pattern.lastIndex = index;
        const match = pattern.exec(input);
        if (match) {
            return new Token(type, match[0], position, ignore);
        }
    }
}
