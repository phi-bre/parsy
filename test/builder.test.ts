import {alternation, optional, parsy, sequence, star} from '../src';

describe('parser', () => {

    const instance = parsy({
        ignore: /[ \n\r]+/,
        terminals: {
            left_brace: '{',
            right_brace: '}',
            left_bracket: '[',
            right_bracket: ']',
            comma: ',',
            colon: ':',
            null: 'null',
            number: /(\d+)?\.?\d+/,
            boolean: /true|false/,
            string: /"(?:[^"\\]|\\.)*"/,
        },
        start: 'value',
        rules: {
            value: alternation('null', 'number', 'boolean', 'string', 'object', 'array'),
            property: sequence('string', 'colon', 'value'),
            object: sequence('left_brace', optional('property', star('comma', 'property')), 'right_brace'),
            array: sequence('left_bracket', optional('value', star('comma', 'value')), 'right_bracket'),
        },
    });

    describe('#parser', () => {
        const input = '[10.5, false, null, [], { "hello": "world" }]';

        it('should not crash', () => {
            expect(instance(input).toString()).toMatchSnapshot();
        });
    });
});
