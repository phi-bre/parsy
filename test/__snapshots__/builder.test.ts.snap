// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser #parser should not crash 1`] = `
"root [
  value [
    array [
      [35mleft_bracket[0m: [
      value [
        [35mnumber[0m: 10.5
      ]
      [35mcomma[0m: ,
      value [
        [35mboolean[0m: false
      ]
      [35mcomma[0m: ,
      value [
        [35mnull[0m: null
      ]
      [35mcomma[0m: ,
      value [
        array [
          [35mleft_bracket[0m: [
          [35mright_bracket[0m: ]
        ]
      ]
      [35mcomma[0m: ,
      value [
        object [
          [35mleft_brace[0m: {
          property [
            [35mstring[0m: \\"hello\\"
            [35mcolon[0m: :
            value [
              [35mstring[0m: \\"world\\"
            ]
          ]
          [35mright_brace[0m: }
        ]
      ]
      [35mright_bracket[0m: ]
    ]
  ]
]
"
`;
